cpp/oneapi/dal/backend/transfer.hpp

```diff
@@ -0,0 +1,91 @@
+/*******************************************************************************
+* Copyright 2020-2021 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#pragma once
+
+#include "oneapi/dal/array.hpp"
+#include "oneapi/dal/backend/memory.hpp"
+
+namespace oneapi::dal::backend {
+
+#ifdef ONEDAL_DATA_PARALLEL
+template <typename T>
+inline std::tuple<array<T>, sycl::event> to_device(sycl::queue& q, const array<T>& ary) {
+    if (!ary.get_count()) {
+        return { ary, sycl::event{} };
+    }
+
+    if (ary.get_queue().has_value()) {
+        // TODO: Consider replacing by exception
+        ONEDAL_ASSERT(ary.get_queue().value() == q,
+                      "We do not support data tranfer between different queues yet");
+
+        const auto ary_alloc = sycl::get_pointer_type(ary.get_data(), q.get_context());
+        if (ary_alloc == sycl::usm::alloc::device) {
+            return { ary, sycl::event{} };
```

Comment:
No memcpy in that case?

Reply:
In addition to that we need to compare devices if we remove assertion above. If we are in the same queues additional check or memcpy is not needed.